### PR TITLE
Refactor cache flow

### PR DESCRIPTION
**Pull request by shimonewman**

### CODE REVIEW

File: src/Database/Database.php [hunk 1]
```diff
@@ -2698,15 +2698,18 @@ public function getDocument(string $collection, string $id, array $queries = [])
...
         $validator = new Authorization(self::PERMISSION_READ);
         $documentSecurity = $collection->getAttribute('documentSecurity', false);
-        $cacheKey = $this->cacheName . '-cache-' . $this->getNamespace() . ':' . $this->adapter->getTenant() . ':' . $collection->getId() . ':' . $id;
 
+        /**
+         * Cache hash keys
+         */
+        $documentCacheBaseKey = $this->cacheName . '-cache-' . $this->getNamespace() . ':' . $this->adapter->getTenant() . ':' . $collection->getId();
+        $documentCacheKey = $documentCacheHash = $documentCacheBaseKey . ':' . $id;
+        $collectionCacheKey = $documentCacheBaseKey . ':collection';
```
Comment (christyjacob4) [inline src/Database/Database.php:2707]:
```suggestion
        $collectionCacheKey = $documentCacheBaseKey;
```
do we need to append `:collection` here? Can we change `$documentCacheBaseKey` to `$documentCacheBaseKey = $this->cacheName . '-cache-' . $this->getNamespace() . ':' . $this->adapter->getTenant() . ':collection:' . $collection->getId()`

File: src/Database/Database.php [hunk 2]
```diff
@@ -4725,21 +4732,33 @@ private function deleteCascade(Document $collection, Document $relatedCollection
      */
     public function purgeCachedCollection(string $collection): bool
     {
-        return $this->cache->purge($this->cacheName . '-cache-' . $this->getNamespace() . ':' . $this->adapter->getTenant() . ':' . $collection . ':*');
+        $collectionKey = $this->cacheName . '-cache-' . $this->getNamespace() . ':' . $this->adapter->getTenant() . ':collection:' . $collection . ':collection';
```
Comment (christyjacob4) [inline src/Database/Database.php:4735]:
```suggestion
        $collectionKey = $this->cacheName . '-cache-' . $this->getNamespace() . ':' . $this->adapter->getTenant() . ':collection:' . $collectionId;
```

File: src/Database/Database.php [hunk 3]
```diff
@@ -4725,21 +4732,33 @@ private function deleteCascade(Document $collection, Document $relatedCollection
      */
     public function purgeCachedCollection(string $collection): bool
```
Comment (christyjacob4) [inline src/Database/Database.php:4733]:
```suggestion
    public function purgeCachedCollection(string $collectionId): bool
```

File: src/Database/Database.php [hunk 4]
```diff
@@ -4725,21 +4732,33 @@ private function deleteCascade(Document $collection, Document $relatedCollection
...
      * Cleans a specific document from cache
+     * And related document reference in the collection cache.
      *
      * @param string $collection
      * @param string $id
      *
      * @return bool
-     * @throws DatabaseException
      */
     public function purgeCachedDocument(string $collection, string $id): bool
```
Comment (christyjacob4) [inline src/Database/Database.php:4753]:
```suggestion
    public function purgeCachedDocument(string $collectionId, string $id): bool
```

File: src/Database/Database.php [hunk 5]
```diff
@@ -4725,21 +4732,33 @@ private function deleteCascade(Document $collection, Document $relatedCollection
...
+        }
+
+        return true;
     }
 
     /**
      * Cleans a specific document from cache
+     * And related document reference in the collection cache.
      *
      * @param string $collection
```
Comment (christyjacob4) [inline src/Database/Database.php:4748]:
```suggestion
     * @param string $collectionId
```

File: src/Database/Database.php [hunk 6]
```diff
@@ -4717,6 +4723,7 @@ private function deleteCascade(Document $collection, Document $relatedCollection
 
     /**
      * Cleans the all the collection's documents from the cache
+     * And the all related cached documents.
      *
      * @param string $collection
```
Comment (christyjacob4) [inline src/Database/Database.php:4728]:
```suggestion
     * @param string $collectionId
```

File: src/Database/Database.php [hunk 7]
```diff
@@ -4725,21 +4732,33 @@ private function deleteCascade(Document $collection, Document $relatedCollection
...
      * @return bool
-     * @throws DatabaseException
      */
     public function purgeCachedDocument(string $collection, string $id): bool
     {
-        return $this->cache->purge($this->cacheName . '-cache-' . $this->getNamespace() . ':' . $this->adapter->getTenant() . ':' . $collection . ':' . $id . ':*');
+        $collectionKey = $this->cacheName . '-cache-' . $this->getNamespace() . ':' . $this->adapter->getTenant() . ':collection:' . $collection;
+        $documentKey =  $collectionKey . ':' . $id;
+
+        $this->cache->purge($collectionKey, $documentKey);
```
Comment (fogelito) [inline src/Database/Database.php:4758]: Do we need to check if purge return bool value? and check for failure?

Comment (shimonewman) [reply src/Database/Database.php:4758]: It's the original behavior as before.
What  do you suggest doing?

File: src/Database/Database.php (hunk 6 above)
Comment (shimonewman) [reply src/Database/Database.php]: Done

Comment (shimonewman) [reply src/Database/Database.php]: Done

File: src/Database/Database.php (hunk 4 above)
Comment (shimonewman) [reply src/Database/Database.php]: done

File: src/Database/Database.php (hunk 2 above)
Comment (shimonewman) [reply src/Database/Database.php]: Done

File: src/Database/Database.php (hunk 3 above)
Comment (shimonewman) [reply src/Database/Database.php]: Done

**Review by christyjacob4 — APPROVED** [review]

**Pull request merged**